src/Processors/Transforms/Streaming/AggregatingHelper.cpp

```diff
@@ -9,111 +9,78 @@ namespace DB
 {
 namespace Streaming
 {
-Chunk AggregatingHelper::convertWithoutKey(AggregatedDataVariants & data, const AggregatingTransformParams & params)
+namespace
 {
-    assert(data.type == AggregatedDataVariants::Type::without_key);
-    return convertToChunk(params.aggregator.prepareBlockAndFillWithoutKey(data, params.final, false, ConvertAction::STREAMING_EMIT));
+Chunk mergeBlocksToChunk(BlocksList && blocks)
+{
+    Chunk merged_chunk;
+    for (auto & block : blocks)
+        merged_chunk.append(DB::convertToChunk(std::move(block)));
+    return merged_chunk;
 }
 
-Chunk AggregatingHelper::mergeAndConvertWithoutKey(ManyAggregatedDataVariants & data, const AggregatingTransformParams & params)
+Chunk convertToChunkImpl(AggregatedDataVariants & data, const AggregatingTransformParams & params, ConvertAction action)
 {
-    auto prepared_data = params.aggregator.prepareVariantsToMerge(data);
-    if (prepared_data->empty())
+    if (data.empty())
         return {};
 
-    assert(prepared_data->at(0)->type == AggregatedDataVariants::Type::without_key);
-    params.aggregator.mergeWithoutKeyDataImpl(*prepared_data, ConvertAction::STREAMING_EMIT);
-    return convertWithoutKey(*prepared_data->at(0), params);
+    auto blocks = params.aggregator.convertToBlocks(data, params.final, action, params.params.max_threads);
+    /// FIXME: When global aggr states was converted two level hash table, the merged chunk may be too large
```

Comment:
Yeah, two level is a good fit to merge one bucket at a time like we have don't for shuffle by aggr

Reply:
Next PR.